### ISSUE TITLE
web api 2 extremely slow

### ISSUE THREAD

**Issue by osigurdson** [mono/mono#13885]
### Current Behavior
Simple asp.net web api 2 example runs more than 50X slower on mono vs .NET. Simple/complete code example below.

### Expected Behavior

Similar performance on mono/Linux vs .NET/Windows.

## On which platforms did you notice this

[ ] macOS
[x ] Linux
[x] Windows

**Version Used**:
 5.18.1.0

## Code Example
~~~~
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net.Http;
using System.Web.Http;
using Owin;
using Microsoft.Owin.Hosting;

namespace slowwebapi2
{
   class Program
   {
      static void Main(string[] args)
      {
         var baseAddress = args.Length == 0 ? "http://+:8000/" : args[0];

         using (WebApp.Start<Startup>(url: baseAddress))
         {
            var client = new HttpClient {BaseAddress = new Uri("http://localhost:8000")};

            var response1 = client.GetAsync("api/values").Result;
            response1.EnsureSuccessStatusCode();
            var sw = Stopwatch.StartNew();
            for (int i = 0; i < 1000; i++)
            {

               var response = client.GetAsync("api/values").Result;
               response.EnsureSuccessStatusCode();
            }

            Console.WriteLine(sw.Elapsed);
            Console.ReadLine();
         }


      }
   }

   class Startup
   {
      public void Configuration(IAppBuilder app)
      {
         // Configure Web API for self-host. 
         HttpConfiguration config = new HttpConfiguration();
         config.Routes.MapHttpRoute(
            name: "DefaultApi",
            routeTemplate: "api/{controller}/{id}",
            defaults: new { id = RouteParameter.Optional }
         );

         app.UseWebApi(config); 
      }
   }

   public class ValuesController : ApiController
   {
      // GET api/values 
      public IEnumerable<string> Get()
      {
         return new string[] { "value1", "value2" };
      }
   }
~~~~
Output on .NET/Windows

00:00:00.7938229

Output on Mono/Linux
00:00:43.3756634

## Packages used
~~~
  <package id="Microsoft.AspNet.WebApi.Client" version="5.2.7" targetFramework="net47" />
  <package id="Microsoft.AspNet.WebApi.Core" version="5.2.7" targetFramework="net47" />
  <package id="Microsoft.AspNet.WebApi.Owin" version="5.2.7" targetFramework="net47" />
  <package id="Microsoft.Owin" version="4.0.1" targetFramework="net47" />
  <package id="Microsoft.Owin.Host.HttpListener" version="4.0.1" targetFramework="net47" />
  <package id="Microsoft.Owin.Hosting" version="4.0.1" targetFramework="net47" />
  <package id="Newtonsoft.Json" version="6.0.4" targetFramework="net47" />
  <package id="Owin" version="1.0" targetFramework="net47" />
~~~

**Comment by osigurdson** [comment]
Added example here: https://github.com/osigurdson/slowwebapi

**Comment by nealef** [comment]
5.18 is exhibiting this behavior but when I run it with binaries built from master the time ends up near 1 second. I am not sure what may have changed in between. I will see if there is anything evident from the git logs.

When run with log profiling under 5.18 the top calls entries are:
```
Method call summary
Total(ms) Self(ms)      Calls Method name
  100521        0          1 (wrapper runtime-invoke) <Module>:runtime_invoke_void_object (object,intptr,intptr,intptr)
  100521      277          1 slowwebapi2.MainClass:Main (string[])
   99564       19       1008 (wrapper runtime-invoke) object:runtime_invoke_void__this__ (object,intptr,intptr,intptr)
   99532    99532       1031 (wrapper managed-to-native) System.Threading.WaitHandle:Wait_internal (intptr*,int,bool,int)
   50155    50155       1012 (wrapper managed-to-native) System.Threading.Monitor:Monitor_wait (object,int)
   47148    47148         26 (wrapper managed-to-native) System.IO.MonoIO:Read (intptr,byte[],int,int,System.IO.MonoIOError&)
   16163      448      32951 (wrapper runtime-invoke) <Module>:runtime_invoke_bool (object,intptr,intptr,intptr)
    6020       23       1974 Microsoft.Owin.Host.HttpListener.OwinHttpListener/<ProcessRequestAsync>d__30:MoveNext ()
    5762       18       2993 Microsoft.Owin.Host.HttpListener.OwinHttpListener/<ProcessRequestsAsync>d__29:MoveNext ()
    5726       18       1019 System.Net.ListenerAsyncResult:InvokeCallback (object)
    5651       40       1979 System.Web.Http.Owin.HttpMessageHandlerAdapter/<InvokeCore>d__20:MoveNext ()
    5637        7       3003 System.Net.Http.HttpMessageInvoker:SendAsync (System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)
```

Whereas the report from master:

```
Method call summary
Total(ms) Self(ms)      Calls Method name
   21121        0          1 (wrapper runtime-invoke) <Module>:runtime_invoke_void_object (object,intptr,intptr,intptr)
   21121       73          1 slowwebapi2.MainClass:Main (string[])
   20417       22       1008 (wrapper runtime-invoke) object:runtime_invoke_void__this__ (object,intptr,intptr,intptr)
   20375    20375          4 (wrapper managed-to-native) System.Threading.WaitHandle:Wait_internal (intptr*,int,bool,int)
   12209      282      19218 (wrapper runtime-invoke) <Module>:runtime_invoke_bool (object,intptr,intptr,intptr)
   10186    10186       1001 (wrapper managed-to-native) System.Threading.Monitor:Monitor_wait (object,int)
    8382     8382         14 (wrapper managed-to-native) System.IO.MonoIO:Read (intptr,byte[],int,int,System.IO.MonoIOError&)
    6926       27       1989 Microsoft.Owin.Host.HttpListener.OwinHttpListener/<ProcessRequestAsync>d__30:MoveNext ()
    6782       30       3008 Microsoft.Owin.Host.HttpListener.OwinHttpListener/<ProcessRequestsAsync>d__29:MoveNext ()
    6768       38       1021 System.Net.ListenerAsyncResult:InvokeCallback (object)
    6607        3       1001 Microsoft.Owin.Host.HttpListener.OwinHttpListener:ProcessRequestAsync (System.Net.HttpListenerContext)
    6602        6       1001 System.Runtime.CompilerServices.AsyncTaskMethodBuilder:Start<Microsoft.Owin.Host.HttpListener.OwinHttpListener/<ProcessRequestAsync>d__30> (Microsoft.Owin.Host.HttpListener.OwinHttpListener/<ProcessRequestAsync>d__30&)
    6542       46       1990 System.Web.Http.Owin.HttpMessageHandlerAdapter/<InvokeCore>d__20:MoveNext ()
    6413       21       3003 System.Net.Http.HttpMessageInvoker:SendAsync (System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)
    6209       16       1001 Microsoft.Owin.Hosting.Utilities.Encapsulate:Invoke (System.Collections.Generic.IDictionary`2<string, object>)
    6141        5       1001 Microsoft.Owin.Infrastructure.OwinMiddlewareTransition:Invoke (System.Collections.Generic.IDictionary`2<string, object>)
    6114        5       1001 System.Web.Http.Owin.HttpMessageHandlerAdapter:Invoke (Microsoft.Owin.IOwinContext)
```

**Comment by osigurdson** [comment]
OK. Good to know.

**Comment by nealef** [comment]
Problem still exists in 5.20.1.19 but not in master. Time for some bisecting.

**Comment by nealef** [comment]
I believe this commit is what fixes the problem:

[2f411750acfd2132f9b4970ef8cf11d9a0f62df9] "Imperial Red": Bring HttpClient from CoreFx (#11906)

**Comment by nealef** [comment]
@directhex Is #11906 scheduled for a release yet? This fix reduces the elapsed time for the test case from 43 seconds to around 1.5 seconds. I am surprised we're the only ones reporting issues with HttpClient.

**Comment by mattzink** [comment]
Our team _just_ upgraded to 5.20, and are heavy users of Web API, so if the offending change can be bisected I would be willing to help develop a fix for 5.x!

**Comment by mattzink** [comment]
Also, if the fix was to HttpClient, is this really a Web API regression, or a client one?

**Comment by osigurdson** [comment]
mattzink,

Did this used to be fast and then regressed (or has it always been slow)?

**Comment by osigurdson** [comment]
Fixed in 6.0.0.313. Performance is now identical to .NET 4.6.1.
